File: src/lib/algorithms/riemersma.ts
// Riemersma dithering (Hilbert curve dithering)
export function riemersmaDithering(
  grayscale: Uint8ClampedArray,
  width: number,
  height: number,
  threshold: number = 128
): ImageData {
  // Create output ImageData
  const outputData = new ImageData(width, height);
  const output = outputData.data;
  
  // Create a copy of the grayscale image
  const buffer = new Float32Array(grayscale);
  
  // Determine the size of the Hilbert curve
  // We'll use a power of 2 that's larger than or equal to the image size
  const maxDimension = Math.max(width, height);
  const order = Math.ceil(Math.log2(maxDimension));
  const curveSize = Math.pow(2, order);
  
  // Generate the Hilbert curve path
  const hilbertPath = generateHilbertCurve(order);
  
  // Set up error diffusion
  const errorBuffer = new Float32Array(3).fill(0); // We'll use a simple 3-tap filter
  
  // Apply dithering along the Hilbert curve
  for (let i = 0; i < hilbertPath.length; i++) {
    const [x, y] = hilbertPath[i];
    
    // Skip points outside the image
    if (x >= width || y >= height) continue;
    
    const idx = y * width + x;
    
    // Add accumulated error to this pixel
    const pixelValue = buffer[idx] + errorBuffer[0];
    
    // Apply threshold
    const newPixel = pixelValue < threshold ? 0 : 255;
    
    // Calculate error
    const error = pixelValue - newPixel;
    
    // Update error buffer (simple shift register)
    errorBuffer[0] = errorBuffer[1] + error * 0.4;
    errorBuffer[1] = errorBuffer[2] + error * 0.3;
    errorBuffer[2] = error * 0.3;
    
    // Set all RGB channels to the result (black or white)
    output[idx * 4] = newPixel;     // R
    output[idx * 4 + 1] = newPixel; // G
    output[idx * 4 + 2] = newPixel; // B
    output[idx * 4 + 3] = 255;      // A (fully opaque)
  }
  
  return outputData;
}

// Generate a Hilbert curve of the given order
function generateHilbertCurve(order: number): [number, number][] {
  const size = Math.pow(2, order);
  const points: [number, number][] = new Array(size * size);
  
  // Initialize with zeros
  for (let i = 0; i < points.length; i++) {
    points[i] = [0, 0];
  }
  
  // Generate the curve
  hilbertCurveHelper(points, 0, 0, size, 0, 0, size - 1, 0, size - 1);
  
  return points;
}

// Helper function to recursively generate the Hilbert curve
function hilbertCurveHelper(
  points: [number, number][],
  index: number,
  x: number,
  y: number,
  xi: number,
  xj: number,
  yi: number,
  yj: number,
  length: number
): number {
  if (length === 1) {
    points[index] = [x, y];
    return index + 1;
  }
  
  length = length / 2;
  
  // Recursively generate the four quadrants
  index = hilbertCurveHelper(
    points, index, 
    x + xi * length, y + yi * length, 
    yi, yj, xi, xj, 
    length
  );
  
  index = hilbertCurveHelper(
    points, index, 
    x + xi * length + yi * length, y + yi * length + yj * length, 
    xi, xj, yi, yj, 
    length
  );
  
  index = hilbertCurveHelper(
    points, index, 
    x + xi * length + yi * length + xi * length, y + yi * length + yj * length + yi * length, 
    xi, xj, yi, yj, 
    length
  );
  
  index = hilbertCurveHelper(
    points, index, 
    x + xi * length + yi * length + xi * length - yj * length, y + yi * length + yj * length + yi * length - xj * length, 
    -yi, -yj, -xi, -xj, 
    length
  );
  
  return index;
} 

File: rust_wasm/src/riemersma.rs
// Hilbert curve generation and Riemersma dithering

// Generate Hilbert curve coordinates for a given order
fn generate_hilbert_curve(order: u32) -> Vec<(usize, usize)> {
    let size = 1usize << order; // 2^order
    let num_points = size * size;
    let mut points = vec![(0, 0); num_points];
    hilbert_curve_recursive(&mut points, 0, 0, 0, 1, 0, 0, 1, order);
    points
}

// Recursive helper for Hilbert curve generation
// Ported from the algorithm described on Wikipedia/various sources
// (x, y) is the starting coordinate of the square
// (xi, yi) and (xj, yj) are vectors defining the orientation of the curve within the square
fn hilbert_curve_recursive(
    points: &mut [(usize, usize)],
    mut index: usize,
    x: usize, y: usize,
    xi: isize, yi: isize,
    xj: isize, yj: isize,
    order: u32,
) -> usize {
    if order == 0 {
        // Base case: place the point
        points[index] = (x, y);
        return index + 1;
    }

    let new_order = order - 1;
    let half_step = 1usize << new_order; // 2^(order-1)

    // Calculate coordinates for the four sub-quadrants
    let x_yi = (x as isize + yi * half_step as isize) as usize;
    let y_yj = (y as isize + yj * half_step as isize) as usize;
    let x_xi = (x as isize + xi * half_step as isize) as usize;
    let y_yi = (y as isize + yi * half_step as isize) as usize;
    let x_xi_yi = (x as isize + xi * half_step as isize + yi * half_step as isize) as usize;
    let y_yi_yj = (y as isize + yi * half_step as isize + yj * half_step as isize) as usize;
    let x_xi_yi_xj = (x as isize + xi * half_step as isize + yi * half_step as isize - xj * half_step as isize) as usize;
    let y_yi_yj_xi = (y as isize + yi * half_step as isize + yj * half_step as isize - xi * half_step as isize) as usize;

    // Recurse into the four sub-quadrants in Hilbert order
    index = hilbert_curve_recursive(points, index, x, y, xj, yj, xi, yi, new_order);
    index = hilbert_curve_recursive(points, index, x_yi, y_yj, xi, yi, xj, yj, new_order);
    index = hilbert_curve_recursive(points, index, x_xi_yi, y_yi_yj, xi, yi, xj, yj, new_order);
    index = hilbert_curve_recursive(points, index, x_xi_yi_xj, y_yi_yj_xi, -xj, -yj, -xi, -yi, new_order);

    index
}

pub fn riemersma_dither(
    input_ptr: *const u8,
    width: usize,
    height: usize,
    threshold: u8,
    output_ptr: *mut u8,
) {
    let size = width * height;
    let input = unsafe { std::slice::from_raw_parts(input_ptr, size) };
    let output = unsafe { std::slice::from_raw_parts_mut(output_ptr, size * 4) }; // RGBA output

    // Create a float buffer of the input
    let buffer: Vec<f32> = input.iter().map(|&p| p as f32).collect();

    // Determine Hilbert curve order needed to cover the image
    let max_dim = width.max(height);
    let order = (max_dim as f32).log2().ceil() as u32;

    // Generate Hilbert curve path
    let hilbert_path = generate_hilbert_curve(order);

    // Error buffer (simple 3-tap)
    let mut error_buffer = [0.0f32; 3];
    let threshold_f = threshold as f32;

    // Apply dithering along the Hilbert curve
    for &(x, y) in &hilbert_path {
        // Check if the point is within the image bounds
        if x < width && y < height {
            let idx = y * width + x;

            // Get pixel value and add accumulated error
            let pixel_value = buffer[idx] + error_buffer[0];

            // Apply threshold
            let new_pixel_value = if pixel_value < threshold_f { 0.0 } else { 255.0 };
            let error = pixel_value - new_pixel_value;

            // Update error buffer (shift and distribute)
            error_buffer[0] = error_buffer[1] + error * 0.4;
            error_buffer[1] = error_buffer[2] + error * 0.3;
            error_buffer[2] = error * 0.3;

            // Write RGBA output pixel
            let base_idx_out = idx * 4;
            let new_pixel_u8 = new_pixel_value as u8;
            output[base_idx_out] = new_pixel_u8;
            output[base_idx_out + 1] = new_pixel_u8;
            output[base_idx_out + 2] = new_pixel_u8;
            output[base_idx_out + 3] = 255; // Alpha
        }
    }
} 